### PR TITLE
Implement MinMax with SIMD

### PR DESCRIPTION
**Pull request by yufenglee**
**Description**: Describe your changes.
Improving the performance of FindMinMax with SIMD

-------------------------------------------------------------------------------
Benchmark                                        Time           CPU Iterations
-------------------------------------------------------------------------------
BM_FindMinMaxPlainLoop/100/real_time            97 ns         96 ns    7287708
BM_FindMinMaxPlainLoop/1000/real_time          961 ns        976 ns     736430
BM_FindMinMaxPlainLoop/10000/real_time        9561 ns       9727 ns      73891
BM_FindMinMaxPlainLoop/20000/real_time       19039 ns      19446 ns      36962
BM_FindMinMaxPlainLoop/40000/real_time       38966 ns      38974 ns      18442
BM_FindMinMaxPlainLoop/80000/real_time       75743 ns      75858 ns       9269
BM_FindMinMaxPlainLoop/98304/real_time       93636 ns      94265 ns       7459
BM_FindMinMaxPlainLoop/160000/real_time     152340 ns     154805 ns       4542

BM_FindMinMaxEigen/100/real_time                26 ns         25 ns   27698690
BM_FindMinMaxEigen/1000/real_time              216 ns        216 ns    3254535
BM_FindMinMaxEigen/10000/real_time            2168 ns       2156 ns     326097
BM_FindMinMaxEigen/20000/real_time            4306 ns       4340 ns     162023
BM_FindMinMaxEigen/40000/real_time            8613 ns       8572 ns      82028
BM_FindMinMaxEigen/80000/real_time           17560 ns      17551 ns      40061
BM_FindMinMaxEigen/98304/real_time           22022 ns      22282 ns      32257
BM_FindMinMaxEigen/160000/real_time          35353 ns      35074 ns      19156

BM_FindMinMaxMlasSSE2/100/real_time             18 ns         18 ns   39980569
BM_FindMinMaxMlasSSE2/1000/real_time           145 ns        143 ns    4815426
BM_FindMinMaxMlasSSE2/10000/real_time         1420 ns       1411 ns     487225
BM_FindMinMaxMlasSSE2/20000/real_time         2854 ns       2826 ns     243295
BM_FindMinMaxMlasSSE2/40000/real_time         5680 ns       5646 ns     124536
BM_FindMinMaxMlasSSE2/80000/real_time        12529 ns      12491 ns      61295
BM_FindMinMaxMlasSSE2/98304/real_time        14007 ns      13863 ns      49594
BM_FindMinMaxMlasSSE2/160000/real_time       22755 ns      22940 ns      30650

BM_FindMinMaxMlasAvx/100/real_time             238 ns        238 ns    2889564
BM_FindMinMaxMlasAvx/1000/real_time            298 ns        299 ns    2355387
BM_FindMinMaxMlasAvx/10000/real_time           987 ns        990 ns     710287
BM_FindMinMaxMlasAvx/20000/real_time          1768 ns       1774 ns     396262
BM_FindMinMaxMlasAvx/40000/real_time          3340 ns       3289 ns     209050
BM_FindMinMaxMlasAvx/80000/real_time          7916 ns       7624 ns      90172
BM_FindMinMaxMlasAvx/98304/real_time          9619 ns       9615 ns      73129
BM_FindMinMaxMlasAvx/160000/real_time        15596 ns      15840 ns      44389

### CODE REVIEW

File: onnxruntime/core/mlas/lib/mlasi.h [hunk 1]
```diff
@@ -1447,6 +1465,28 @@ MlasReduceMaximumFloat32x4(MLAS_FLOAT32X4 Vector)
 #endif
 }
 
+MLAS_FORCEINLINE
+float
+MlasReduceMinimumFloat32x4(MLAS_FLOAT32X4 Vector) {
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/mlasi.h]: brace on next line

File: onnxruntime/core/mlas/lib/quantize.cpp [hunk 2]
```diff
@@ -422,3 +422,37 @@ Return Value:
 }
 
 #endif
+
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/quantize.cpp:425]: can you make this look more MlasComputeTanh? The comment style is not the same, ) is not done the same,

File: onnxruntime/core/mlas/lib/quantize.cpp [hunk 3]
```diff
@@ -422,3 +422,37 @@ Return Value:
...
+       N - Supplies the number of elements to process.
+
+   --*/
+{
+#if defined(MLAS_TARGET_AMD64)
+    MlasPlatform.MinMaxF32Kernel(
+#else
+    MlasMinMaxF32Kernel(
+#endif
+        Input, min, max, N);
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/quantize.cpp:457]: can this be merged into the previous lines like MlasComputeTanh?

File: onnxruntime/core/mlas/lib/mlasi.h [hunk 4]
```diff
@@ -448,6 +448,18 @@ float
...
+(MLASCALL MLAS_MIN_MAX_FLOAT_KERNEL)(
+    const float* Input,
+    float* Min,
+    float* Max,
+    size_t N
+    );
+
+typedef MLAS_MIN_MAX_FLOAT_KERNEL* PMLAS_MIN_MAX_FLOAT_KERNEL;
+
+
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/mlasi.h:462]: remove space

File: onnxruntime/core/mlas/lib/intrinsics/basic/min_max_elements.cpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+    This module implements the logic to find min and max elements with basic extension instructions.
+
+--*/
+
+#include "../mlasi.h"
+
+void
+MLASCALL
+MlasMinMaxF32Kernel(
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/intrinsics/basic/min_max_elements.cpp:21]: I would have just called this MlasReduceMinimumMaximumF32Kernel and put this after the MlasReduceMaximumF32Kernel in compute.cpp and not added a new file.

File: onnxruntime/core/mlas/lib/intrinsics/basic/min_max_elements.cpp [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    float* max,
+    size_t N)
+{
+    if (N <= 0) {
+        *min = 0.0f;
+        *max = 0.0f;
+        return;
+    }
+
+    *min = *max = Input[0];
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/intrinsics/basic/min_max_elements.cpp:33]: Using the style from MlasReduceMaximumF32Kernel I would have just initialized a local to the min/max values and not needed to read Input unneccessarily here. The N <=0 case above wouldn't be necessary.

File: onnxruntime/core/mlas/lib/intrinsics/basic/min_max_elements.cpp [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+    if (N >= 4) {
+
+        MLAS_FLOAT32X4 MaximumVector0 = MlasBroadcastFloat32x4(*Input);
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/intrinsics/basic/min_max_elements.cpp:37]: Why is this different from MlasReduceMaximumF32Kernel?

File: onnxruntime/core/mlas/lib/intrinsics/basic/min_max_elements.cpp [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+            MinimumVector0 = MlasMinimumFloat32x4(MinimumVector0, MinimumVector1);
+            MinimumVector2 = MlasMinimumFloat32x4(MinimumVector2, MinimumVector3);
+            MinimumVector0 = MlasMinimumFloat32x4(MinimumVector0, MinimumVector2);
+        }
+
+        while (N >= 4) {
+
+            MLAS_FLOAT32X4 InputVector0 = MlasLoadFloat32x4(Input);
+            MaximumVector0 = MlasMaximumFloat32x4(MaximumVector0, InputVector0);
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/intrinsics/basic/min_max_elements.cpp:83]: add space?

File: onnxruntime/core/mlas/lib/mlasi.h [hunk 9]
```diff
@@ -448,6 +448,18 @@ float
 
 typedef MLAS_REDUCE_MAXIMUM_FLOAT_KERNEL* PMLAS_REDUCE_MAXIMUM_FLOAT_KERNEL;
 
+typedef
+void
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/mlasi.h:452]: Can this be called MLAS_REDUCE_MINIMUM_MAXIMUM_FLOAT_KERNEL? Or at least MLAS_REDUCE_MIN_MAX_FLOAT_KERNEL?

File: onnxruntime/core/mlas/lib/intrinsics/avx2/min_max_elements.cpp [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+Licensed under the MIT License.
+
+Module Name:
+
+    min_max_elements.cpp
+
+Abstract:
+
+    This module implements the logic to find min and max elements with avx2 instructions.
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/intrinsics/avx2/min_max_elements.cpp:13]: everything here appears to be AVX only: AVX has minps/maxps, so the kernel could be enabled for AVX, like MlasReduceMaximumF32KernelAvx.

File: onnxruntime/core/mlas/lib/intrinsics/avx2/min_max_elements.cpp [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+
+void
+MLASCALL
+MlasMinMaxF32KernelAvx2(
+    const float* Input,
+    float* min,
+    float* max,
+    size_t N)
+{
+    if (N <= 0) {
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/intrinsics/avx2/min_max_elements.cpp:27]: same general feedback as the generic version

File: onnxruntime/core/mlas/lib/intrinsics/avx2/min_max_elements.cpp [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+
+            __m256 InputVector0 = _mm256_loadu_ps(Input);
+            MaximumVector0 = _mm256_max_ps(MaximumVector0, InputVector0);
+            MinimumVector0 = _mm256_min_ps(MinimumVector0, InputVector0);
+
+            Input += 8;
+            N -= 8;
+        }
+
+        float min_buf[8], max_buf[8];
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/intrinsics/avx2/min_max_elements.cpp:90]: can this instead just do a vector extract of the upper half, then do a min/max of that, then use MlasReduceMaximumFloat32x4/MlasReduceMinimumFloat32x4?

File: onnxruntime/core/mlas/lib/intrinsics/avx/min_max_elements.cpp [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+
+Licensed under the MIT License.
+
+Module Name:
+
+    min_max_elements.cpp
+
+Abstract:
+
+    This module implements the logic to find min and max elements with avx2 instructions.
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/intrinsics/avx/min_max_elements.cpp:13]: avx2->AVX

File: onnxruntime/core/mlas/lib/quantize.cpp [hunk 14]
```diff
@@ -422,3 +422,36 @@ Return Value:
...
 
 #endif
+
+void
+MLASCALL
+MlasFindMinMaxElement(
+    const float* Input,
+    float* min,
+    float* max,
+    size_t N )
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/quantize.cpp]: ) on new line

File: onnxruntime/core/mlas/lib/quantize.cpp [hunk 15]
```diff
@@ -422,3 +422,36 @@ Return Value:
...
+   /*++
+
+   Routine Description:
+
+       This routine implements the generic kernel to find the minimum and maximum value of
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/quantize.cpp]: The comment should be updated: this isn't the generic kernel here

File: onnxruntime/core/mlas/lib/quantize.cpp [hunk 16]
```diff
@@ -422,3 +422,36 @@ Return Value:
 }
 
 #endif
+
+void
+MLASCALL
+MlasFindMinMaxElement(
+    const float* Input,
+    float* min,
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/quantize.cpp]: Parameter names should be Minimum/Maximum to match existing style.

File: onnxruntime/test/mlas/unittest.cpp [hunk 17]
```diff
@@ -2363,6 +2363,59 @@ class MlasComputeExpTest : public MlasTestBase
...
+
+        float diff_max = std::fabs(max - max_ref);
+        if (diff_max > epsilon) {
+            printf("maximum difference: %.8f %.8f\n", max, max_ref);
+        }
+    }
+
+public:
+    void
+        ExecuteShort(
```
Comment (tracysh) [inline onnxruntime/test/mlas/unittest.cpp:2409]: spacing here is off

File: onnxruntime/test/mlas/unittest.cpp [hunk 18]
```diff
@@ -2363,6 +2363,59 @@ class MlasComputeExpTest : public MlasTestBase
     }
 };
 
+class MlasFindMinMaxElementsTest : public MlasTestBase
+{
+private:
+    MatrixGuardBuffer<float> BufferInput;
+
+    void
+        Test(
```
Comment (tracysh) [inline onnxruntime/test/mlas/unittest.cpp:2372]: spacing here is off

File: onnxruntime/core/mlas/lib/compute.cpp [hunk 19]
```diff
@@ -561,6 +561,88 @@ Return Value:
...
+            N -= 4;
+        }
+
+        *min = MlasReduceMinimumFloat32x4(MinimumVector0);
+        *max = MlasReduceMaximumFloat32x4(MaximumVector0);
+    }
+
+    while (N > 0) {
+
+        *max = std::max(*max, *Input);
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/compute.cpp:638]: Can you look at the generated code to see if this is doing a load/store on each iteration here? I would have just had a local float min/max that you do the reduction to, same as MlasReduceMaximumF32Kernel. min/max don't need to be written to until the very end.

File: onnxruntime/core/mlas/lib/intrinsics/avx/min_max_elements.cpp [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+
+Licensed under the MIT License.
+
+Module Name:
+
+    min_max_elements.cpp
+
+Abstract:
+
+    This module implements the logic to find min and max elements with avx instructions.
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/intrinsics/avx/min_max_elements.cpp:13]: Please uppercase the term AVX here.

File: onnxruntime/core/mlas/lib/compute.cpp (hunk 19 above)
Comment (yufenglee) [reply onnxruntime/core/mlas/lib/compute.cpp:638]: the compiler doesn't optimize it. it load/store for each iteration.

**Pull request merged**